Fix PyDictIterator's segfault(for #159)

kngwyu opened the pull request:
Fix #159 
Thanks @fdoyon for debugging!

konstin commented:
Thank you and a big :+1: for both of you! I'll add some style comments because if we add the test as example people will copy it's style so we should be careful not to promote bad style.

konstin commented on examples/rustapi_module/src/dict_iter.rs:
```diff
@@ -0,0 +1,42 @@
+            );
+        }
+
+        match seen == self.expected {
```
This should be if (afaik clippy also complains about that)

konstin commented on examples/rustapi_module/src/dict_iter.rs:
```diff
@@ -0,0 +1,42 @@
+use pyo3::exceptions as exc;
+use pyo3::types::PyDict;
+
+#[pymodinit(_test_dict)]
```
I'd like to move all examples to module names without underscores

konstin commented on examples/rustapi_module/src/dict_iter.rs:
```diff
@@ -0,0 +1,42 @@
+use pyo3::prelude::*;
+
+use pyo3::exceptions as exc;
```
No need to rename here

konstin commented on examples/rustapi_module/src/dict_iter.rs:
```diff
@@ -0,0 +1,42 @@
+impl DictSize {
+    #[new]
+    fn __new__(obj: &PyRawObject, expected: u32) -> PyResult<()> {
+        obj.init(|_t| DictSize { expected })
```
This will need `||` since PyToken was removed.

konstin commented on examples/rustapi_module/tests/test_dict_iter.py:
```diff
@@ -0,0 +1,12 @@
+)
+def test_size(size):
+    d = {}
+    for i in range(0,size):
```
`range(size)`

konstin commented on src/types/dict.rs:
```diff
@@ -148,13 +148,19 @@ impl PyDict {
+pub struct PyDictIterator<'dict> {
+    dict: PyObject,
     pos: isize,
+    __marker: PhantomData<&'dict ()>,
```
Is there a special reason not to use `Python` as marker here. Also imho we don't need the underscores since the fields are private anyway

kngwyu commented:
Ah but looks test fails after rebasing.
Please wait for a while.

konstin commented on src/types/dict.rs:
```diff
@@ -166,7 +172,7 @@ impl<'a> Iterator for PyDictIterator<'a> {
             let mut value: *mut ffi::PyObject = mem::uninitialized();
             if ffi::PyDict_Next(self.dict.as_ptr(), &mut self.pos, &mut key, &mut value) != 0 {
-                let py = self.dict.py();
+                let py = Python::assume_gil_acquired();
```
I prefer the old version here since it makes sure that this method isn't changed into some gil-less variant one day and this becomes essentially UB

kngwyu replied:
`PyObject` doesn't implement `PyObjectWithGIL`.
We need to change this line if we use `PyObject`

konstin replied:
ok that's fine; Could you add that as a comment for future reference?

kngwyu replied:
Now I changed `PyDictIterator` to have `Python`.
Still comment is needed?

konstin replied:
No, that's clear now

konstin commented on src/types/dict.rs:
```diff
@@ -175,6 +181,12 @@ impl<'a> Iterator for PyDictIterator<'a> {
 
+impl<'py> Drop for PyDictIterator<'py> {
+    fn drop(&mut self) {
+        unsafe { ffi::Py_DECREF(self.dict.as_ptr()) }
```
This confuses me: Where have we called INCREF so we can DECREF here?

kngwyu replied:
In PyDict::iter.
`to_object` calls `from_borrowed_object`.

kngwyu commented:
Thanks for your review!
I fixed review points and added
```rust
 impl<'py> Drop for PyDictIterator<'py> {
    fn drop(&mut self) {
        unsafe { ffi::Py_DECREF(self.dict.as_ptr()) }
    }
}
```
.
Since we call INCREF for PyObject when getting it from `&PyDict` so I think it's correct, but not sure.

konstin commented:
Afaik that's already done by 

https://github.com/PyO3/pyo3/blob/aed0466915527f15656b8dbc343bd0d00c4e7682/src/object.rs#L315-L322

kngwyu commented:
Ah, thanks, I missed it.

fdoyon commented:
Thanks gents, I had hoped to contribute some code but it seems I am missing some understanding of the memory model. :+1:

kngwyu commented:
@fdoyon 
The key point is `&PyDict` is a pointer to `*mut ffi::PyObject` and it points an element in ReleasePool.
```rust
pub unsafe fn register_owned(_py: Python, obj: *mut ffi::PyObject) -> &PyObjectRef {
    let pool: &'static mut ReleasePool = &mut *POOL;
    pool.owned.push(obj);
    &*(&pool.owned[pool.owned.len() - 1] as *const *mut ffi::PyObject as *const PyObjectRef)
}
```

So we can avoid segfault by having `*mut ffi::PyObject` directly.
But, as you commented, it's still a big problem. This kind of bug can happens anywhere in our codebase.

kngwyu commented:
Made a separate issue #271
